### PR TITLE
Clean up clippy allows

### LINKED ISSUE #1127: Way too many `clippy::allow`
It seems we've got in the habit of doing `clippy::allow` at many points that clippy complains about something. I count ~20 places. I think we should fix most these and refrain from doing it in the future (don't merge new PRs with `clippy::allow` in them).

### PR DESCRIPTION
**Pull request by ValuedMammal**
closes #1127 

There are several instances in the code where we allow clippy lints that would otherwise be flagged during regular checks. It would be preferable to minimize the number of "clippy allow" attributes either by fixing the affected areas or setting a specific configuration in `clippy.toml`. In cases where we have to allow a particular lint, it should be documented why the lint doesn't apply.

For context see https://github.com/bitcoindevkit/bdk/issues/1127#issuecomment-1784256647 as well as the commit message details.

One area I'm unsure of is whether `Box`ing a large error in 4fc2216 is the right approach. Logically it makes sense to avoid allocating a needlessly heavy `Result`, but I haven't studied the implications or tradeoffs of such a change.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

### CODE REVIEW

**Comment by ValuedMammal** [comment]
The first attempt failed CI due to `clippy::iter_nth_zero`. Of course, there's no reason to think the lint could be disregarded in the first place.. I believe I used the wrong invocation of `cargo clippy` on the command line, allowing it to slip by. I suppose the nix experience should eventually help prevent silly mistakes.

After taking another look at spk_iter.rs, I think it's fine to allow the exception (as it was originally), since we're specifically testing `SpkIterator`'s custom implementation of `nth`. An alternative would be to write the tests in a way that doesn't require calling `nth(0)`, but for now this seems like a trivial case not worth worrying about.

File: nursery/tmp_plan/src/requirements.rs [hunk 1]
```diff
@@ -117,6 +117,7 @@ impl From<bip32::Error> for SigningError {
 impl std::error::Error for SigningError {}
 
 impl RequiredSignatures<DescriptorPublicKey> {
+    #[allow(clippy::too_many_arguments)]
```
Comment (LLFourn) [inline nursery/tmp_plan/src/requirements.rs]: feel free to just #[allow(clippy::all)]  in this module's lib.rs

File: example-crates/example_cli/src/lib.rs [hunk 2]
```diff
@@ -183,18 +183,17 @@ impl core::fmt::Display for Keychain {
     }
 }
 
-#[allow(clippy::type_complexity)]
+// Alias changeset info in `create_tx`
+pub type ChangeSet = (keychain::ChangeSet<Keychain>, (Keychain, u32));
```
Comment (danielabrozzoni) [inline example-crates/example_cli/src/lib.rs]: Instead of defining this alias, can we change the return parameter of create_tx to return a structure? It can contain:
- transaction
- keychain_changeset
- change_keychain
- change_index

I think this would make the code a bit clearer, what do you think?

File: crates/esplora/src/async_ext.rs [hunk 3]
```diff
@@ -6,11 +6,17 @@ use bdk_chain::{
...
+use esplora_client::TxStatus;
 use futures::{stream::FuturesOrdered, TryStreamExt};
 
 use crate::{anchor_from_status, ASSUME_FINAL_DEPTH};
 
+/// `Box` a large [`esplora_client::Error`]
+type Error = Box<esplora_client::Error>;
+
+/// Alias `Result` to use our `Error`
+type Result<T, E = Error> = core::result::Result<T, E>;
```
Comment (danielabrozzoni) [inline crates/esplora/src/async_ext.rs]: Why is this line needed?

**Review by danielabrozzoni — COMMENTED** [review]
Overall it looks really good, thanks a lot!
A few questions/suggestions:

File: crates/esplora/src/async_ext.rs (hunk 3 above)
Comment (ValuedMammal) [reply crates/esplora/src/async_ext.rs]: You're right it's not needed, let's get rid of it. The thought was that it allows omitting the `Err` variant in a function signature similar to what you get from `anyhow::Result`, but I neglected to actually remove them. In this case, it's probably better to be explicit and keep `Error` in the fn signatures.

File: example-crates/example_cli/src/lib.rs (hunk 2 above)
Comment (ValuedMammal) [reply example-crates/example_cli/src/lib.rs]: Yes, I like that. However, since three of these fields are either all `Some` or all `None`, it might make sense to have a struct just for `change_info` and continue to return a tuple of `(transaction, change_info)` from `create_tx`

For example, here I introduce `CreateTxChange` and make its members match the names of the existing logic in fn `handle_commands`, does that make sense?

```
--- a/example-crates/example_cli/src/lib.rs
+++ b/example-crates/example_cli/src/lib.rs
@@ -181,21 +181,24 @@ impl core::fmt::Display for Keychain {
             Keychain::Internal => write!(f, "internal"),
         }
     }
 }
 
-// Alias changeset info in `create_tx`
-pub type ChangeSet = (keychain::ChangeSet<Keychain>, (Keychain, u32));
+pub struct CreateTxChange {
+    pub index_changeset: keychain::ChangeSet<Keychain>,
+    pub change_keychain: Keychain,
+    pub index: u32,
+}
 
 pub fn create_tx<A: Anchor, O: ChainOracle>(
     graph: &mut KeychainTxGraph<A>,
     chain: &O,
     keymap: &HashMap<DescriptorPublicKey, DescriptorSecretKey>,
     cs_algorithm: CoinSelectionAlgo,
     address: Address,
     value: u64,
-) -> anyhow::Result<(Transaction, Option<ChangeSet>)>
+) -> anyhow::Result<(Transaction, Option<CreateTxChange>)>
 where
     O::Error: std::error::Error + Send + Sync + 'static,
 {
     let mut changeset = keychain::ChangeSet::default();
 
@@ -385,11 +388,15 @@ where
             }
         }
     }
 
     let change_info = if selection_meta.drain_value.is_some() {
-        Some((changeset, (internal_keychain, change_index)))
+        Some(CreateTxChange {
+            index_changeset: changeset,
+            change_keychain: internal_keychain,
+            index: change_index,
+        })
     } else {
         None
     };
 
     Ok((transaction, change_info))
@@ -593,11 +600,16 @@ where
                 let graph = &mut *graph.lock().unwrap();
                 // take mutable ref to construct tx -- it is only open for a short time while building it.
                 let (tx, change_info) =
                     create_tx(graph, chain, keymap, coin_select, address, value)?;
 
-                if let Some((index_changeset, (change_keychain, index))) = change_info {
+                if let Some(CreateTxChange {
+                    index_changeset,
+                    change_keychain,
+                    index,
+                }) = change_info
+                {
                     // We must first persist to disk the fact that we've got a new address from the
                     // change keychain so future scans will find the tx we're about to broadcast.
                     // If we're unable to persist this, then we don't want to broadcast.
                     {
                         let db = &mut *db.lock().unwrap();
```

Comment (danielabrozzoni) [reply example-crates/example_cli/src/lib.rs]: Yes, sounds good!

File: crates/bdk/src/wallet/coin_selection.rs [hunk 4]
```diff
@@ -531,10 +517,19 @@ impl BranchAndBoundCoinSelection {
         optional_utxos.sort_unstable_by_key(|a| a.effective_value);
         optional_utxos.reverse();
 
+        let mut curr_value = required_utxos
```
Comment (danielabrozzoni) [inline crates/bdk/src/wallet/coin_selection.rs]: I don't think it's worth the hassle to refactor the cs module, since it's going away soon. Also see https://github.com/bitcoindevkit/bdk/issues/1127#issuecomment-1784313256

**Comment by ValuedMammal** [comment]
rebasing now. dropped all the changes to `coin_selection.rs`, and made `ref(tmp_plan)` the last commit

File: crates/bdk/src/wallet/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs:1]: I guess I'm on the fence regarding `preselect_utxos`. @LLFourn would you prefer we left this untouched so it's clear at the call site exactly what information we're passing?

**Comment by ValuedMammal** [comment]
Reminder to bump msrv in clippy.toml
```
warning: the MSRV in `clippy.toml` and `Cargo.toml` differ; using `1.57.0` from `clippy.toml`
```

**Comment by ValuedMammal** [comment]
Changes in most recent push:

- re-refactored `Psbt::get_utxo_for` to use a match block. I think it conveys the same thing in a more concise way. [compare](https://github.com/bitcoindevkit/bdk/compare/6cdc7f1a95f2a26f65b87ead00cc57b328f18f51..6107725756493ff5f0afa94c25c11d0a598933a1#diff-504dcd23159cb3b14ff38d9c2148749e8a213d292c346c55035f36c0aae7ccd6)

File: nursery/tmp_plan/src/lib.rs [hunk 5]
```diff
@@ -12,7 +12,7 @@
 //! doing coin selection. Furthermore it provides which subset of those keys and hash pre-images you
 //! will actually need as well as what locktime or sequence number you need to set.
 //!
-//! Once you've obstained signatures, hash pre-images etc required by the plan, it can create a
+//! Once you've obtained signatures, hash pre-images etc required by the plan, it can create a
```
Comment (LLFourn) [inline nursery/tmp_plan/src/lib.rs]: I'd prefer we didn't touch this crate and just allowed all.

**Comment by LLFourn** [comment]
Please revert all changes to the tmp plan crate and just blanket allow everything. Other than that this looks good to go.

**Comment by ValuedMammal** [comment]
@LLFourn appreciate your input, thanks for clarifying.

File: example-crates/example_cli/src/lib.rs [hunk 6]
```diff
@@ -643,17 +651,19 @@ where
     }
 }
 
-#[allow(clippy::type_complexity)]
-pub fn init<CS: clap::Subcommand, S: clap::Args, C>(
-    db_magic: &[u8],
-    db_default_path: &str,
-) -> anyhow::Result<(
+// Alias the `Result` of `init`
+pub type InitialState<CS, S, C> = (
```
Comment (LLFourn) [inline example-crates/example_cli/src/lib.rs]: This could actually be made into a struct where each field had a nice comment if you felt up to it.

**Review by LLFourn — APPROVED** [review]
ACK c4f121109198df3a3f96b507fe76a7f14146a135

Thanks @ValuedMammal!

File: example-crates/example_cli/src/lib.rs (hunk 6 above)
Comment (ValuedMammal) [reply example-crates/example_cli/src/lib.rs]: Would you rename this?

**Comment by ValuedMammal** [comment]
Is anyone opposed to making the type `InitialState` in `example_cli` into a struct with comments?

```rust
/// The initial state returned from [`init`].
pub struct Init<CS: clap::Subcommand, S: clap::Args, C> {
    /// Arguments parsed by the cli.
    pub args: Args<CS, S>,
    /// Descriptor keymap.
    pub keymap: KeyMap,
    /// Keychain-txout index.
    pub index: KeychainTxOutIndex<Keychain>,
    /// Persistence backend.
    pub db: Mutex<Database<C>>,
    /// Initial changeset.
    pub init_changeset: C,
}
```

**Comment by danielabrozzoni** [comment]
@ValuedMammal go for it!

File: crates/bdk/src/psbt/mod.rs [hunk 7]
```diff
@@ -35,24 +35,20 @@ pub trait PsbtUtils {
...
     fn get_utxo_for(&self, input_index: usize) -> Option<TxOut> {
         let tx = &self.unsigned_tx;
 
-        if input_index >= tx.input.len() {
-            return None;
-        }
+        let input = match self.inputs.get(input_index) {
+            Some(input) => input,
+            None => return None,
+        };
```
Comment (evanlinjin) [inline crates/bdk/src/psbt/mod.rs]: Nit:

```suggestion
        let input = self.inputs.get(input_index)?;
```

File: example-crates/example_cli/src/lib.rs [hunk 8]
```diff
@@ -388,43 +388,46 @@ where
...
     } else {
         None
     };
 
     Ok((transaction, change_info))
 }
 
-#[allow(clippy::type_complexity)]
+// Alias the elements of `Result` of `planned_utxos`
+pub type Plan<K, A> = (bdk_tmp_plan::Plan<K>, FullTxOut<A>);
```
Comment (evanlinjin) [inline example-crates/example_cli/src/lib.rs]: Nit: can we rename this to `PlannedUTXO`? Thanks!

**Review by evanlinjin — COMMENTED** [review]
Looking good! Just a couple of NITs

File: crates/bdk/src/psbt/mod.rs (hunk 7 above)
Comment (ValuedMammal) [reply crates/bdk/src/psbt/mod.rs]: indeed, this is better.

File: example-crates/example_cli/src/lib.rs (hunk 8 above)
Comment (ValuedMammal) [reply example-crates/example_cli/src/lib.rs]: I hope you don't mind I used `PlannedUtxo`.

**Comment by ValuedMammal** [comment]
- Rebased onto f099b42005b5310cee51fa2eaa29f94d76753bd7
- Addressed comments by @evanlinjin 

Thank you to all the reviewers! @danielabrozzoni @LLFourn

**Review by evanlinjin — APPROVED** [review]
ACK 1c15cb2f9169bb08a4128c183eeeca80fff97ab7

File: example-crates/example_cli/src/lib.rs (hunk 8 above)
Comment (evanlinjin) [reply example-crates/example_cli/src/lib.rs]: Woops sorry I was in Golang world these two weeks.
